ripc-protocol-tcp/src/main/java/io/ripc/protocol/tcp/TcpServer.java

```diff
@@ -0,0 +1,88 @@
+package io.ripc.protocol.tcp;
+
+import io.ripc.internal.Publishers;
+import org.reactivestreams.Publisher;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public abstract class TcpServer<R, W> {
+
+    @SuppressWarnings("rawtypes")
+    private LazyTcpHandler lazyHandler;
+    protected final TcpHandler<R, W> thisHandler;
+    protected final AtomicBoolean started;
+
+    protected TcpServer() {
+        lazyHandler = new LazyTcpHandler<>();
+        thisHandler = null;
+        this.started = new AtomicBoolean();
+    }
+
+    protected TcpServer(TcpHandler<R, W> handler) {
+        thisHandler = handler;
+        this.started = new AtomicBoolean();
+    }
+
+    public <RR, WW> TcpServer<RR, WW> intercept(TcpInterceptor<R, W, RR, WW> interceptor) {
+        if (null == thisHandler) {
+            @SuppressWarnings("unchecked")
+            TcpServer<RR, WW> toReturn = newServer(interceptor.intercept(lazyHandler));
+            toReturn.lazyHandler = lazyHandler;
+            return toReturn;
+        } else {
+            TcpServer<RR, WW> toReturn = newServer(interceptor.intercept(thisHandler));
+            toReturn.lazyHandler = lazyHandler;
+            return toReturn;
+        }
+    }
+
+    public final TcpServer<R, W> start(TcpHandler<R, W> handler) {
+        if (!started.compareAndSet(false, true)) {
+            throw new IllegalStateException("Server already started");
+        }
+
+        if (null == thisHandler) {
+            doStart(handler);
+        } else {
+            lazyHandler.start(handler);
+            doStart(thisHandler);
+        }
+        return this;
+    }
+
+    public final void startAndAwait(TcpHandler<R, W> handler) {
```

Comment:
Should we be mixing blocking semantics into a fundamentally non-blocking server? It's likely that few "real" situations would use this anyway since there would be other means of keeping the JVM alive (a la Spring Boot, which would definitely not use an await() method. Can't we accomplish this via helpers that take a `Publisher` and block until `Subscriber.onComplete`?